### DISCUSSION TITLE
Content-Type = application/json; utf-8

### DISCUSSION THREAD

**Discussion by fabiojsantosdev** [adonisjs/core#4150] [Help] [answered]
Good afternoon.

I'm implementing a payment by pix, where the Bank has a service that needs a route to notify if payment has been made. I need some data that arrives in the body of this request. The problem is that they Content-Type = application/json; utf-8, this Content-Type is causing request.body() to return an empty object. Could anyone help with this problem?

**Comment by thetutlage** [comment]
Can you share a screenshot of the request you are making? Along with headers and the request body?

**Reply by fabiojsantosdev** [reply]
> Can you share a screenshot of the request you are making? Along with headers and the request body?

Headers
![image](https://github.com/adonisjs/core/assets/92104539/7ca884eb-1045-4ba1-825a-6878844e4996)

Json

![image](https://github.com/adonisjs/core/assets/92104539/aac393c8-f6d0-4f6d-87c9-c0575cc80254)


Create a Middleware to use in the route with it I can get the data from the body. The problem is that when I use it in the routine it is closing the request before the return of the model and controller.

Below is the code:

import { HttpContextContract } from '@ioc:Adonis/Core/HttpContext'

export default class RawBodyParser {
   public async handle({ request }: HttpContextContract, next: () => Promise<void>) {
     let data = ''
     request.request.on('data', (chunk: any) => {

       date += chunk
     })

      request.request.on('end', () => {
       request.body = JSON.parse(data)

        next()
     })

   }
}

**Reply by RomainLanz** [reply]
Why are you reading the HTTP request chunk by hand instead of relying on [Adonis function](https://docs.adonisjs.com/guides/request) or even [Node.js helpers](https://nodejs.org/api/webstreams.html#utility-consumers)?

**Comment by McSneaky** [comment, marked as answer]
Copy-pasting it from my Discord response

Won't accept, coz it's invalid header
According to standard, only acceptable header is `application/json` without anything extra
But also it's quite widespread to use `application/json; charset=utf-8`
`charset` needs to be defined in most APIs if you want it to be parsed all good
Just slapping `utf-8` in the end doesn't follow even any recommendations
According to standard it can be `utf-8`, `utf-16` or `utf-32` and it should be detected by analyzing first 4 bytes of incoming data, not based on content type header

Standard itself: https://datatracker.ietf.org/doc/html/rfc4627

I'd send off email to bank to fix it on their side
Meanwhile, I'll see if can figure out fix for you

```ts
import type { HttpContextContract } from '@ioc:Adonis/Core/HttpContext'

export default class Customjson {
  public async handle({request, }: HttpContextContract, next: () => Promise<void>) {
    // Check for non-standard JSON content-type header
    if (request.header('content-type') === 'application/json; utf-8') {
      let rawBuffer = await request.request.read()

      let rawBodyString = rawBuffer.toString('utf-8')

      let parsedBody = JSON.parse(rawBodyString)

      request.setInitialBody(parsedBody)
    }

    await next()
  }
}
```
That works as a middleware to use

**Comment by fabiojsantosdev** [comment]
Thanks for helping resolve the issue. It could have this treatment directly in AdonisJS.

**Reply by McSneaky** [reply]
I'm on a fence with supporting it

On one hand it would be good to support it, but then on the other hand, supporting non-standard stuff is kinda pain. In future might need to add support for `application/json, utf8` and `application/json: utf-8` etc. Becomes huge burden in the end
But if that becomes common enuf, then would be good to add them to framework itself

**Comment by fabiojsantosdev** [comment]
It could have support, as I had this problem in the integration of BRADESCO's PIX API. That according to them application/json, utf8 is not standard. Their application is developed by IBM. I got to test the integration with AspNet.Core and I had no problem accessing the Body's content, but as my application was built on top of AdonisJs, I was looking for a solution so I didn't have to create a microservice with another language/Framework.
